src/WaveClip.cpp

```diff
@@ -1789,3 +1712,233 @@ const wxString& WaveClip::GetName() const
+{
+    SetSequenceStartTime(time - mTrimLeft);
+}
+
+double WaveClip::GetPlayEndTime() const
+{
+    auto numSamples = mSequence->GetNumSamples();
+
+    double maxLen = GetSequenceStartTime() + ((numSamples + mAppendBufferLen).as_double()) / mRate - mTrimRight;
+    // JS: calculated value is not the length;
+    // it is a maximum value and can be negative; no clipping to 0
+
+    return maxLen;
+}
+
+sampleCount WaveClip::GetPlayStartSample() const
+{
+    return TimeToSamples(GetPlayStartTime());
+}
+
+sampleCount WaveClip::GetPlayEndSample() const
+{
+    return GetPlayStartSample() + GetPlaySamplesCount();
+}
+
+sampleCount WaveClip::GetPlaySamplesCount() const
+{
+    return mSequence->GetNumSamples()
+       - TimeToSamples(mTrimRight) - TimeToSamples(mTrimLeft);
+}
+
+void WaveClip::SetTrimLeft(double trim)
+{
+    mTrimLeft = std::max(.0, trim);
+}
+
+double WaveClip::GetTrimLeft() const noexcept
+{
+    return mTrimLeft;
+}
+
+void WaveClip::SetTrimRight(double trim)
+{
+    mTrimRight = std::max(.0, trim);
+}
+
+double WaveClip::GetTrimRight() const noexcept
+{
+    return mTrimRight;
+}
+
+void WaveClip::TrimLeft(double deltaTime)
+{
+    mTrimLeft += deltaTime;
+}
+
+void WaveClip::TrimRight(double deltaTime)
+{
+    mTrimRight += deltaTime;
+}
+
+void WaveClip::TrimLeftTo(double to)
+{
+    mTrimLeft = std::clamp(to, GetSequenceStartTime(), GetPlayEndTime()) - GetSequenceStartTime();
+}
+
+void WaveClip::TrimRightTo(double to)
+{
+    mTrimRight = GetSequenceEndTime() - std::clamp(to, GetPlayStartTime(), GetSequenceEndTime());
+}
+
+double WaveClip::GetSequenceStartTime() const noexcept
+{
+    // JS: mSequenceOffset is the minimum value and it is returned; no clipping to 0
+    return mSequenceOffset;
+}
+
+void WaveClip::SetSequenceStartTime(double startTime)
+{
+    mSequenceOffset = startTime;
+    mEnvelope->SetOffset(startTime);
+}
+
+double WaveClip::GetSequenceEndTime() const
+{
+    auto numSamples = mSequence->GetNumSamples();
+
+    double maxLen = GetSequenceStartTime() + (numSamples + mAppendBufferLen).as_double() / mRate;
+    // JS: calculated value is not the length;
+    // it is a maximum value and can be negative; no clipping to 0
+
+    return maxLen;
+}
+
+sampleCount WaveClip::GetSequenceStartSample() const
+{
+    return TimeToSamples(mSequenceOffset);
+}
+
+sampleCount WaveClip::GetSequenceEndSample() const
+{
+    return GetSequenceStartSample() + mSequence->GetNumSamples();
+}
+
+void WaveClip::Offset(double delta) noexcept
+{
+    SetSequenceStartTime(GetSequenceStartTime() + delta);
+}
+
+// Bug 2288 allowed overlapping clips.
+// This was a classic fencepost error.
+// We are within the clip if start < t <= end.
+// Note that BeforeClip and AfterClip must be consistent 
+// with this definition.
+bool WaveClip::WithinPlayRegion(double t) const
+{
+    auto ts = (sampleCount)floor(t * mRate + 0.5);
+    return ts > GetPlayStartSample() && ts < GetPlayEndSample() + mAppendBufferLen;
```

Comment:
This copies the comment and the logic of `WaveClip::WithinClip()` but it looks strange.  The fix for 2288 changed `<=` to `<`.  But should it have changed `>` to `>=` instead?

Reply:
There is also another comment saying:
` One and only one of the following is true for a given t (unless the clip has zero length -- then BeforeClip() and AfterClip() can both be true). Within() is true if the time is substantially within the clip`. 
I think this is more about the definition of 'within'. Changing the implementation will require a careful review of the usage of each function. And I think that changing '<=' to '<' was anyway correct, because 'GetPlayEndSample()' is the index of the sample following the last one, so 'within' includes the last sample, but does not include the first sample.